lib/chef/provider/package/macospkg.rb

```diff
@@ -0,0 +1,93 @@
+#
+# Author:: Joshua Timberman (<jtimberman@chef.io>)
+# Author:: William Theaker (<william.theaker+chef@gusto.com>)
+# Copyright:: Copyright (c) Chef Software Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require_relative "../provider/package"
+require_relative "package"
+
+class Chef
+  class Provider
+    class MacosPkg < Chef::Provider::Package
+      provides :macos_pkg
+
+      def load_current_resource
+        @current_resource = Chef::Resource::Package.new(new_resource.name)
+        current_resource.package_name(new_resource.package_name)
+        current_resource.version(get_current_version)
+        logger.trace("#{new_resource} current package version: #{current_resource.version}") if current_resource.version
+
+        current_resource
+      end
+
+      def get_current_version
+        shell_out("pkgutil --pkg-info '#{new_resource.package_id}'").stdout.to_s[/version: (.*)/, 1]
+      end
+
+      def define_resource_requirements
+        requirements.assert(:install) do |a|
+          a.assertion { new_resource.source || new_resource.file }
+          a.failure_message Chef::Exceptions::Package, "Must provide either a file or source property for #{new_resource.package_name} macos_pkg resource."
+        end
+      end
+
+      def upgrade_package(name, version)
+        forget_package_receipt if current_resource.version
+        install_package(name, version)
+      end
+
+      def install_package(name, version)
+        download_pkg if new_resource.source
+        shell_out("installer -pkg #{pkg_file} -target #{new_resource.target}")
+      end
+
+      def forget_package_receipt
+        shell_out("pkgutil --forget '#{new_resource.package_id}'")
+      end
+
+      def remove_package(name, version)
```

Comment:
Yeah this is pretty decent, why not just call that from `upgrade_package` instead of `forget_package_receipt` ?

Reply:
Agreed, macOS treats upgrades differently than installs and if a package maintainer does their job correctly, the bom (bill of materials) in the package will "do the right thing". Breaking that logic seems non-idea.